### PR TITLE
Move work out of HttpConnectionPool.ReturnConnection lock

### PR DESCRIPTION
**Pull request by stephentoub**
Small tweaks (doesn't move the needle much), but some of the work being done inside the lock can be done before or after it instead.

cc: @geoffkizer, @mikeharder

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 1]
```diff
@@ -848,69 +847,70 @@ public void DecrementConnectionCount()
...
+
+                    // If there's someone waiting for a connection and this one's still valid, simply transfer this one to them rather than pooling it.
+                    // Note that while we checked connection lifetime above, we don't check idle timeout, as even if idle timeout
+                    // is zero, we consider a connection that's just handed from one use to another to never actually be idle.
+                    bool receivedUnexpectedData = false;
+                    if (_waitersTail != null)
                     {
-                        ConnectionWaiter waiter = DequeueWaiter();
-                        waiter._cancellationTokenRegistration.Dispose();
+                        receivedUnexpectedData = connection.EnsureReadAheadAndPollRead();
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: This call is likely the most expensive thing we do under the lock.

I wonder if this is even necessary at all. We've just finished a request on this connection, and as part of this we've ensured that we didn't read any more data than we expected. So why do we check again here?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: > and as part of this we've ensured that we didn't read any more data than we expected.

Where did we already check that, and specifically where did we check that the server didn't send more data than it should have?

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: That's true. Though these things should only occur with an ill-behaved server.

If you think we should keep the check, then ideally we'd do it outside the lock. Possibly the waiter could do the test after the waiter is completed. Doesn't have to be in this PR, of course. I'm mucking with all this waiter stuff in my PR anyway, I could look at how to do this.

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Also consider that we have retry logic that already handles the race conditions. This isn't perfect either, but it does sort of mitigate the possible issues here.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Ok, let me think about it.

**Pull request merged**